src/differentiation/compute_jacobian_ad.jl

```diff
@@ -79,23 +93,37 @@ function forwarddiff_color_jacobian!(J::AbstractMatrix{<:Number},
     p = jac_cache.p
     color = jac_cache.color
 
-    # TODO: Should compute on each p[i] and decompress
-    t .= Dual{typeof(f)}.(x, p)
-    f(fx, t)
-
-    if J isa SparseMatrixCSC
-        rows_index, cols_index, val = findnz(J)
-        for color_i in 1:maximum(color)
-            dx .= partials.(fx,color_i)
-            for i in 1:length(cols_index)
-                if color[cols_index[i]]==color_i
-                    J[rows_index[i],cols_index[i]] = dx[rows_index[i]]
+    color_i = 1
+    chunksize = getsize(default_chunk_size(maximum(color)))
+
+    for i in 1:length(p)
+        partial_i = p[i]
+        t .= Dual{typeof(f)}.(x, partial_i)
+        f(fx,t)
+
+        if J isa SparseMatrixCSC
+            rows_index, cols_index, val = findnz(J)
+            for j in 1:chunksize
+                dx .= partials.(fx, j)
+                for k in 1:length(cols_index)
+                    if color[cols_index[k]] == color_i
+                        J[rows_index[k], cols_index[k]] = dx[rows_index[k]]
+                    end
+                end
+                color_i += 1
+                if color_i == maximum(color) + 1
+                    color_i = 1
```

Comment:
this doesn't seem to be used?

Reply:
That's because the chunk size defined here 
~~~
function default_chunk_size(maxcolor)
    if maxcolor < DEFAULT_CHUNK_THRESHOLD
        Val(maxcolor)
    else
        Val(DEFAULT_CHUNK_THRESHOLD)
    end
end
~~~
ensures that the size of a chunk is always less than or equal to the number of colors used. If we decide to use chunk sizes that are larger than the maximum value of color, we'll have to make sure `color_i` wraps around instead of going past the largest value in coloring vector.

Reply:
Oh okay nevermind my bad. That made no sense.